File: yogi-cpp/test/yogi/context_test.cc
/*
 * This file is part of the Yogi Framework
 * https://github.com/yohummus/yogi-framework.
 *
 * Copyright (c) 2020 Johannes Bergmann.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <test/common.h>

#include <chrono>
#include <regex>
#include <thread>
using namespace std::chrono_literals;

class ContextTest : public Test {
 public:
  yogi::ContextPtr context_ = create_context();
};

TEST_F(ContextTest, Poll) {
  MOCK_ContextPoll([](void* context, int* count) {
    EXPECT_NE(context, nullptr);
    EXPECT_NE(count, nullptr);
    *count = 5;
    return YOGI_OK;
  });

  EXPECT_EQ(context_->poll(), 5);
}

TEST_F(ContextTest, PollError) {
  MOCK_ContextPoll([](void*, int*) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(context_->poll(), yogi::FailureException);
}

TEST_F(ContextTest, PollOne) {
  MOCK_ContextPollOne([](void* context, int* count) {
    EXPECT_NE(context, nullptr);
    EXPECT_NE(count, nullptr);
    *count = 1;
    return YOGI_OK;
  });

  EXPECT_EQ(context_->poll_one(), 1);
}

TEST_F(ContextTest, PollOneError) {
  MOCK_ContextPollOne([](void*, int*) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(context_->poll_one(), yogi::FailureException);
}

TEST_F(ContextTest, Run) {
  MOCK_ContextRun([](void* context, int* count, long long duration) {
    EXPECT_NE(context, nullptr);
    EXPECT_NE(count, nullptr);
    EXPECT_EQ(duration, 1000000ll);
    *count = 5;
    return YOGI_OK;
  });

  EXPECT_EQ(context_->run(1ms), 5);
}

TEST_F(ContextTest, RunError) {
  MOCK_ContextRun([](void*, int*, long long duration) {
    EXPECT_EQ(-1, duration);
    return YOGI_ERR_UNKNOWN;
  });

  EXPECT_THROW(context_->run(), yogi::FailureException);
}

TEST_F(ContextTest, RunOne) {
  MOCK_ContextRunOne([](void* context, int* count, long long duration) {
    EXPECT_NE(context, nullptr);
    EXPECT_NE(count, nullptr);
    EXPECT_EQ(duration, 1000000ll);
    *count = 1;
    return YOGI_OK;
  });

  EXPECT_EQ(context_->run_one(1ms), 1);
}

TEST_F(ContextTest, RunOneError) {
  MOCK_ContextRunOne([](void*, int*, long long duration) {
    EXPECT_EQ(-1, duration);
    return YOGI_ERR_UNKNOWN;
  });

  EXPECT_THROW(context_->run_one(), yogi::FailureException);
}

TEST_F(ContextTest, RunInBackground) {
  MOCK_ContextRunInBackground([](void* context) {
    EXPECT_NE(context, nullptr);
    return YOGI_OK;
  });

  EXPECT_NO_THROW(context_->run_in_background());
}

TEST_F(ContextTest, RunInBackgroundError) {
  MOCK_ContextRunInBackground([](void*) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(context_->run_in_background(), yogi::FailureException);
}

TEST_F(ContextTest, Stop) {
  MOCK_ContextStop([](void* context) {
    EXPECT_NE(context, nullptr);
    return YOGI_OK;
  });

  EXPECT_NO_THROW(context_->stop());
}

TEST_F(ContextTest, StopError) {
  MOCK_ContextStop([](void*) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(context_->stop(), yogi::FailureException);
}

TEST_F(ContextTest, WaitForRunning) {
  MOCK_ContextWaitForRunning([](void* context, long long duration) {
    EXPECT_NE(context, nullptr);
    EXPECT_EQ(duration, 1000000ll);
    return YOGI_OK;
  });

  EXPECT_TRUE(context_->wait_for_running(1ms));
}

TEST_F(ContextTest, WaitForRunningTimeout) {
  MOCK_ContextWaitForRunning([](void*, long long) { return YOGI_ERR_TIMEOUT; });

  EXPECT_FALSE(context_->wait_for_running(1ms));
}

TEST_F(ContextTest, WaitForRunningError) {
  MOCK_ContextWaitForRunning([](void*, long long duration) {
    EXPECT_EQ(-1, duration);
    return YOGI_ERR_UNKNOWN;
  });

  EXPECT_THROW(context_->wait_for_running(), yogi::FailureException);
}

TEST_F(ContextTest, WaitForStopped) {
  MOCK_ContextWaitForStopped([](void* context, long long duration) {
    EXPECT_NE(context, nullptr);
    EXPECT_EQ(duration, 1000000ll);
    return YOGI_OK;
  });

  EXPECT_TRUE(context_->wait_for_stopped(1ms));
}

TEST_F(ContextTest, WaitForStoppedTimeout) {
  MOCK_ContextWaitForStopped([](void*, long long) { return YOGI_ERR_TIMEOUT; });

  EXPECT_FALSE(context_->wait_for_stopped(1ms));
}

TEST_F(ContextTest, WaitForStoppedError) {
  MOCK_ContextWaitForStopped([](void*, long long duration) {
    EXPECT_EQ(-1, duration);
    return YOGI_ERR_UNKNOWN;
  });

  EXPECT_THROW(context_->wait_for_stopped(), yogi::FailureException);
}

TEST_F(ContextTest, Post) {
  MOCK_ContextPost([](void* context, void (*fn)(void* userarg), void* userarg) {
    EXPECT_EQ(context, kPointer);
    EXPECT_NE(fn, nullptr);
    EXPECT_NE(userarg, nullptr);
    fn(userarg);
    return YOGI_OK;
  });

  bool called = false;
  auto fn     = [&] { called = true; };
  context_->post(fn);
  EXPECT_TRUE(called);
}

TEST_F(ContextTest, PostError) {
  MOCK_ContextPost([](void*, void (*)(void*), void*) { return YOGI_ERR_UNKNOWN; });
  EXPECT_THROW(context_->post({}), FailureException);
}

File: yogi-dotnet/test/ContextTest.cs
/*
 * This file is part of the Yogi Framework
 * https://github.com/yohummus/yogi-framework.
 *
 * Copyright (c) 2020 Johannes Bergmann.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

using System;
using Xunit;

namespace test
{
    public class ContextTest : TestCase
    {
        Yogi.Context context = CreateContext();

        [Fact]
        public void Create()
        {
            MOCK_ContextCreate((ref IntPtr context) =>
            {
                context = pointer;
                return (int)Yogi.ErrorCode.Ok;
            });

            new Yogi.Context();
        }

        [Fact]
        public void CreateError()
        {
            MOCK_ContextCreate((ref IntPtr context) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                new Yogi.Context();
            });
        }

        [Fact]
        public void Poll()
        {
            MOCK_ContextPoll((IntPtr context, ref int count) =>
            {
                Assert.Equal(pointer, context);
                count = 5;
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.Equal(5, context.Poll());
        }

        [Fact]
        public void PollError()
        {
            MOCK_ContextPoll((IntPtr context, ref int count) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                context.Poll();
            });
        }

        [Fact]
        public void PollOne()
        {
            MOCK_ContextPollOne((IntPtr context, ref int count) =>
            {
                Assert.Equal(pointer, context);
                count = 1;
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.Equal(1, context.PollOne());
        }

        [Fact]
        public void PollOneError()
        {
            MOCK_ContextPollOne((IntPtr context, ref int count) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                context.PollOne();
            });
        }

        [Fact]
        public void Run()
        {
            MOCK_ContextRun((IntPtr context, ref int count, long duration) =>
            {
                Assert.Equal(pointer, context);
                Assert.Equal(1000000L, duration);
                count = 5;
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.Equal(5, context.Run(Yogi.Duration.FromMilliseconds(1)));
        }

        [Fact]
        public void RunError()
        {
            MOCK_ContextRun((IntPtr context, ref int count, long duration) =>
            {
                Assert.Equal(-1L, duration);
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                context.Run();
            });
        }

        [Fact]
        public void RunOne()
        {
            MOCK_ContextRunOne((IntPtr context, ref int count, long duration) =>
            {
                Assert.Equal(pointer, context);
                Assert.Equal(1000000L, duration);
                count = 1;
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.Equal(1, context.RunOne(Yogi.Duration.FromMilliseconds(1)));
        }

        [Fact]
        public void RunOneError()
        {
            MOCK_ContextRunOne((IntPtr context, ref int count, long duration) =>
            {
                Assert.Equal(-1L, duration);
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                context.RunOne();
            });
        }

        [Fact]
        public void RunInBackground()
        {
            MOCK_ContextRunInBackground((IntPtr context) =>
            {
                Assert.Equal(pointer, context);
                return (int)Yogi.ErrorCode.Ok;
            });

            context.RunInBackground();
        }

        [Fact]
        public void RunInBackgroundError()
        {
            MOCK_ContextRunInBackground((IntPtr context) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                context.RunInBackground();
            });
        }

        [Fact]
        public void Stop()
        {
            MOCK_ContextStop((IntPtr context) =>
            {
                Assert.Equal(pointer, context);
                return (int)Yogi.ErrorCode.Ok;
            });

            context.Stop();
        }

        [Fact]
        public void StopError()
        {
            MOCK_ContextStop((IntPtr context) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                context.Stop();
            });
        }

        [Fact]
        public void WaitForRunning()
        {
            MOCK_ContextWaitForRunning((IntPtr context, long duration) =>
            {
                Assert.Equal(pointer, context);
                Assert.Equal(1000000L, duration);
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.True(context.WaitForRunning(Yogi.Duration.FromMilliseconds(1)));
        }

        [Fact]
        public void WaitForRunningTimeout()
        {
            MOCK_ContextWaitForRunning((IntPtr context, long duration) =>
            {
                return (int)Yogi.ErrorCode.Timeout;
            });

            Assert.False(context.WaitForRunning(Yogi.Duration.FromMilliseconds(1)));
        }

        [Fact]
        public void WaitForRunningError()
        {
            MOCK_ContextWaitForRunning((IntPtr context, long duration) =>
            {
                Assert.Equal(-1L, duration);
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                context.WaitForRunning();
            });
        }

        [Fact]
        public void WaitForStopped()
        {
            MOCK_ContextWaitForStopped((IntPtr context, long duration) =>
            {
                Assert.Equal(pointer, context);
                Assert.Equal(1000000L, duration);
                return (int)Yogi.ErrorCode.Ok;
            });

            Assert.True(context.WaitForStopped(Yogi.Duration.FromMilliseconds(1)));
        }

        [Fact]
        public void WaitForStoppedTimeout()
        {
            MOCK_ContextWaitForStopped((IntPtr context, long duration) =>
            {
                return (int)Yogi.ErrorCode.Timeout;
            });

            Assert.False(context.WaitForStopped(Yogi.Duration.FromMilliseconds(1)));
        }

        [Fact]
        public void WaitForStoppedError()
        {
            MOCK_ContextWaitForStopped((IntPtr context, long duration) =>
            {
                Assert.Equal(-1L, duration);
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                context.WaitForStopped();
            });
        }

        [Fact]
        public void Post()
        {
            MOCK_ContextPost((IntPtr context, ContextPostFnDelegate fn, IntPtr userarg) =>
            {
                Assert.Equal(pointer, context);
                fn(userarg);
                return (int)Yogi.ErrorCode.Ok;
            });

            bool called = false;
            Action handlerFn = () => { called = true; };
            context.Post(handlerFn);
            Assert.True(called);
        }

        [Fact]
        public void PostError()
        {
            MOCK_ContextPost((IntPtr context, ContextPostFnDelegate fn, IntPtr userarg) =>
            {
                return (int)Yogi.ErrorCode.Unknown;
            });

            Assert.ThrowsAny<Yogi.FailureException>(() =>
            {
                context.Post(() => { });
            });
        }
    }
}
